File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.php
<?php
$CASE_SENSITIVE = false;

function compareStrings ($text1, $text2){
	$out = false;
	
	global $CASE_SENSITIVE;
	if ($CASE_SENSITIVE){
		$text1 = strtoupper($text1);
		$text2 = strtoupper($text2);
	}
	
	if ($text1 === "" && $text2 === ""){
		$out = true;
		return ($out);
	}
	
	if (strlen($text1) != strlen($text2)){
		return ($out);
	}
	
	for ($i = 0; $i < strlen($text1); $i++){
		if ($i == 0){
			if ($text1[$i] == $text2[$i]){
				$out = true;
			} else {
				$out = false;
			}
		} else {
			$out = $out && ($text1[$i] == $text2[$i]);
		}
	}
	
	return ($out);
}

function firstOccurence ($needle, $haystack){
	$lenNeedle = strlen($needle);
	$lenHaystack = strlen($haystack);
	$out = -1;
	
	if ($lenNeedle > $lenHaystack){
		$out = -1;
	} else if ($lenNeedle == $lenHaystack){
		$found = compareStrings($needle, $haystack);
		$out = $found ? 0 : -1;
	} else {
		$lenSearchable = $lenHaystack - $lenNeedle + 1;
		for ($i = 0; $i < $lenSearchable; $i++){
			if (compareStrings($needle, substr($haystack, $i, ($i + $lenNeedle)))){ //this is the part of the string
				$out = $i;
				break;
			}
		}
	}
	
	return ($out);
}

echo((firstOccurence("ode", "leetcode")) . "\n\n\n");
?>

File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.py
CASE_SENSITIVE = False

def compare_strings (text1, text2):
	out = False
	
	global CASE_SENSITIVE
	if CASE_SENSITIVE:
		text1 = text1.upper()
		text2 = text2.upper()
	
	if (text1 == "" and text2 == ""):
		out = True
		return out
	
	if len(text1) != len(text2):
		return out
	
	for i in range (len(text1)):
		if (i == 0):
			if (text1[i] == text2[i]):
				out = True
			else:
				out = False
		else:
			out = out and (text1[i] == text2[i])
	
	return out
	
def first_occurence (needle, haystack):
	len_needle = len(needle)
	len_haystack = len(haystack)
	out = -1
	
	if len_needle > len_haystack:
		out = -1
	elif len_needle == len_haystack:
		found = compare_strings(needle, haystack)
		out = 0 if found else -1
	else:
		len_searchable = len_haystack - len_needle + 1
		for i in range(len_searchable):
			#print (needle + "\t" + haystack[i: (i + len_needle)])
			if (compare_strings(needle, haystack[i: (i + len_needle)])):
				out = i
				break
	
	return out
	

print(first_occurence( haystack = "leetcode", needle = "ode"))
